monix-tail/shared/src/main/scala/monix/tail/batches/Batch.scala

```diff
@@ -163,19 +163,19 @@ object Batch {
     * @param offset $paramArrayOffset
     * @param length $paramArrayLength
     */
-  def fromArray[A : ClassTag](array: Array[A], offset: Int, length: Int): ArrayBatch[A] =
-    new ArrayBatch[A](array, offset, length)
+  def fromArray[A](array: Array[A], offset: Int, length: Int): ArrayBatch[A] = {
+    val tp = ClassTag[A](array.getClass.getComponentType)
+    new ArrayBatch[A](array, offset, length)(tp)
+  }
 
   /** $fromAnyArrayDesc
     *
     * @param array $paramArray
     * @param offset $paramArrayOffset
     * @param length $paramArrayLength
     */
-  def fromAnyArray[A](array: Array[_], offset: Int, length: Int): ArrayBatch[A] = {
-    val ref = new ArrayBatch[Any](array.asInstanceOf[Array[Any]], offset, length, arrayAnyBuilder)
-    ref.asInstanceOf[ArrayBatch[A]]
-  }
+  def fromAnyArray[A](array: Array[_], offset: Int, length: Int): ArrayBatch[A] =
+    fromArray(array, offset, length).asInstanceOf[ArrayBatch[A]]
```

Comment:
Is this method even needed anymore? Judging by the doc it has been introduced exactly for the cases where you don't have a `ClassTag`, and is not really type-safe, i.e. this compiles:

    Batch.fromAnyArray[Int](Array("A string!"), 0, 1)

Reply:
Due to Array's reification it might be needed. This is because Java allows boxing of primitives into `AnyRef`, so you can do this:

```scala
val array = new Array[Any](10)
array(0) = 1001
```

And we have to do this unfortunately when describing a `map[B](f: A => B)` operation in `ArrayCursor` for example. What happens is that we build a new `Array[Any]`, iterate the source and initialize it, but at the end you have an `Array[Any]` on your hands and it's not something you can workaround unless you require a `ClassTag[B]`.

And then manual casting doesn't work either, due to reification:

```
scala> array.asInstanceOf[Array[Int]]
java.lang.ClassCastException: [Ljava.lang.Object; cannot be cast to [I
  ... 28 elided
```

So unfortunately I think it's still needed in order to work with `Array[Any]`, due to reification.

I hope generics reification never makes it in Java 🙂

Reply:
Sure, you cannot cast an array. But you can cast the returned `ArrayCursor` itself.

Reply:
Ah, right 🙂 Will try it out and if so, we'll get rid of it.